File: 프로그래머스/2/154539. 뒤에 있는 큰 수 찾기/뒤에 있는 큰 수 찾기.py
def solution(numbers):
    answer = [-1] * len(numbers)
    stack = []
    stack.append(0)
    
    for i in range(1, len(numbers)):
        while stack and numbers[stack[-1]] < numbers[i]:
            popped = stack.pop()
            answer[popped] = numbers[i]
        stack.append(i)
    return answer

'''
O(N^2) 하면 10^12
10^8보다 한참

모노스택 써서 푸는거같다.
모노스택의 탑에는 가장 큰게 와있어야하나?
or 가장 작은게 와있어야 하나?
    직관적으로는 가장 큰게 와있어야 할거같은데 생각을 해보자
    
'''

File: 프로그래머스/2/154539. 뒤에 있는 큰 수 찾기/뒤에 있는 큰 수 찾기.java
import java.util.*;
class Solution {
    public int[] solution(int[] numbers) {
        Stack<Integer> stack = new Stack<>();//인덱스
        //stack.push(numbers[0]);
        int[] ans = new int[numbers.length];
        Arrays.fill(ans, -1);//초기값
        for (int i = 0; i < numbers.length; i++) {
            while (!stack.isEmpty() && numbers[stack.peek()] < numbers[i]) {
                //i가 계속 크면 갱신
                int poppedIdx = stack.pop();
                ans[poppedIdx] = numbers[i]; //답은 숫자..idx넣는이유는 ans때문
            }//다 처리되면 이제 stack은 더큰값이 있고 ..제일작은 number[i].
            stack.push(i);
        }
        return ans;
    }
}
/**
단조스택이네.
10^7이므로 O(N)이 맥스임.
*/